### PR TITLE
Adding Event client for APIM

### PR DESCRIPTION
**Pull request by shilmyhasan**
Includes APIM Event client to publish sample data to apim related siddhi applications

### CODE REVIEW

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        String host = args[1];
+        String port = args[2];
+        int sslPort = Integer.parseInt(port) + 100;
+        String username = args[3];
+        String password = args[4];
+        String numberOfEventsStr = args[5];
+        int numberOfEvents = Integer.parseInt(numberOfEventsStr);
+
+        try {
+            log.info("Starting WSO2 Event Client");
```
Comment (dilini-muthumala) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]:
```suggestion
            log.info("Starting WSO2 APIM Sample Event Client");
```

**Review by dilini-muthumala — CHANGES_REQUESTED** [review]

**Review by dilini-muthumala — APPROVED** [review]

File: modules/samples/sample-clients/apim-event-client/build.xml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            <include name="org.wso2.carbon.secvault_5.0.*.jar"/>
+            <include name="org.wso2.carbon.utils_2.0.*.jar"/>
+            <include name="org.wso2.carbon.databridge.agent_6.0.*.jar"/>
+            <include name="org.wso2.carbon.databridge.commons.thrift_6.0.*.jar"/>
+            <include name="org.wso2.carbon.databridge.commons.binary_6.0.*.jar"/>
+            <include name="org.wso2.carbon.databridge.commons_6.0.*.jar"/>
+            <include name="org.ops4j.pax.logging.pax-logging-log4j2_*.jar"/>
+            <include name="org.ops4j.pax.logging.pax-logging-api_*.jar"/>
+        </fileset>
+
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/build.xml]: we can remove this additional space.

File: modules/samples/sample-clients/apim-event-client/pom.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        <dependency>
+            <groupId>org.wso2.carbon.analytics-common</groupId>
+            <artifactId>org.wso2.carbon.databridge.commons.binary</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.analytics-common</groupId>
+            <artifactId>org.wso2.carbon.databridge.agent</artifactId>
+        </dependency>
+    </dependencies>
+
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/pom.xml]: please remove additional space.

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                requestEvent.setPayloadData(requestData);
+                dataPublisher.publish(event);
+                dataPublisher.publish(faultEvent);
+                dataPublisher.publish(requestEvent);
+            }
+            try {
+                Thread.sleep(5000);
+            } catch (InterruptedException e) {
+                log.error(e);
+            }
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: please add meaning full error message.

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+            dataPublisher.shutdown();
+            log.info("Events published successfully");
+
+        } catch (Throwable e) {
+            log.error(e);
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: please add meaning full error message.

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ */
+
+package org.wso2.carbon.sample.apimevent;
+
+import java.io.File;
+
+/**
+ * Datapublisher Util Implementataion.
+ */
+public class DataPublisherUtil {
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: please add method level comments.

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                    File.separator + "main" + File.separator + "resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("test" + File.separator + "resources");
+        }
+        String trustStore = filePath.getAbsolutePath();
+        System.setProperty("javax.net.ssl.trustStore", trustStore + File.separator + "client-truststore.jks");
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: can get the JKS file names from configurations. If it's file names get changed, there could be an issue, right?

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                    File.separator + "main" + File.separator + "resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("test" + File.separator + "resources");
+        }
+        String keyStore = filePath.getAbsolutePath();
+        System.setProperty("Security.KeyStore.Location", keyStore + File.separator + "wso2carbon.jks");
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: hardcoded Keystore name and credentials. Explain the reason.

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                    "sample-clients" + File.separator + "apimevent-client" + File.separator + "src" +
+                    File.separator + "main" + File.separator + "resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("test" + File.separator + "resources");
+        }
+        return filePath.getAbsolutePath() + File.separator + fileName;
```
Comment (ruks) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: There are some string used multiple times. Can we make them class level constants?

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java (hunk 1 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: fixed

File: modules/samples/sample-clients/apim-event-client/build.xml (hunk 2 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/build.xml]: fixed in 1f69fe0038ac7ea895a7fae7e7714add8fc82281

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java (hunk 4 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: Fixed in 1f69fe0038ac7ea895a7fae7e7714add8fc82281

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java (hunk 5 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: Fixed in 1f69fe0038ac7ea895a7fae7e7714add8fc82281

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java (hunk 6 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: Fixed in 1f69fe0038ac7ea895a7fae7e7714add8fc82281

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java (hunk 7 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: Its an internal test client and JKS files are loaded internally from the resouces. The file will not change

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java (hunk 8 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: Its an internal test client and JKS files are loaded internally from the resouces.

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java (hunk 9 above)
Comment (shilmyhasan) [reply modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: Fixed in 1f69fe0038ac7ea895a7fae7e7714add8fc82281

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+        throttledObj.applicationId = applicationId[index % 3];
+        throttledObj.applicationName = applicationName[index % 3];
+        throttledObj.subscriber = subscriber[index % 2];
+        throttledObj.throttledOutReason = throttledOutReason[index % 2];
+        throttledObj.gatewayType = "g1";
+        throttledObj.throttledOutTimestamp = new Timestamp(System.currentTimeMillis()).getTime();
+        throttledObj.hostname = "localhost";
+
+        return (new Object[]{
+                throttledObj.username,
```
Comment (HiranyaKavishani) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: we can reduce the lines (L145 - L158)

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        faultObj.hostname = "localhost";
+        faultObj.applicationId = applicationId[index % applicationName.length];
+        faultObj.applicationName = applicationName[index % applicationName.length];
+        faultObj.protocol = "https";
+        faultObj.errorCode = "403";
+        faultObj.errorMessage = "notfound";
+        faultObj.requestTimestamp = new Timestamp(System.currentTimeMillis()).getTime();
+
+        return (new Object[]{
+              faultObj.applicationConsumerKey,
```
Comment (HiranyaKavishani) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: same here please reduce the lines

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        requestObj.throttlingLatency = (long) 2;
+        requestObj.requestMedLat = (long) 2;
+        requestObj.responseMedLat = (long) 2;
+        requestObj.backendLatency = (long) 2;
+        requestObj.otherLatency = (long) 2;
+        requestObj.gatewayType = "SYNAPSE";
+        requestObj.label = "SYNAPSE";
+
+        return (new Object[]{
+           requestObj.applicationConsumerKey,
```
Comment (HiranyaKavishani) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/Client.java]: same here

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        File filePath = new File("src" + File.separator + "main" + File.separator + "resources");
+        if (!filePath.exists()) {
+            filePath = new File("modules" + File.separator + "samples" + File.separator +
+                    "sample-clients" + File.separator + "apimevent-client" + File.separator + "src" +
+                    File.separator + "main" + File.separator + "resources");
+        }
+        if (!filePath.exists()) {
+            filePath = new File("resources");
+        }
+        if (!filePath.exists()) {
```
Comment (HiranyaKavishani) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/DataPublisherUtil.java]: Why we need L35 and L38 both? What is the scenario that comes to second condition?

File: modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/FaultDTO.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+ * under the License.
+ */
+package org.wso2.carbon.sample.apimevent;
+
+/**
+ * Fault DTO contains attributes of Fault Stream
+ */
+public class FaultDTO {
+
+    String applicationConsumerKey;
```
Comment (HiranyaKavishani) [inline modules/samples/sample-clients/apim-event-client/src/main/java/org/wso2/carbon/sample/apimevent/FaultDTO.java]: We can reduce the lines at here

**Pull request merged**